[nnc] tensorexpr for quantized::add

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70193
* #70189
* __->__ #70188

Differential Revision: [D33238093](https://our.internmc.facebook.com/intern/diff/D33238093)

IvanKobzarev committed:
[nnc] tensorexpr for quantized::add

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238093).

IvanKobzarev committed:
Update on "[nnc] tensorexpr for quantized::add"


Differential Revision: [D33238093](https://our.internmc.facebook.com/intern/diff/D33238093)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238093).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/operators/quantization.cpp:
```diff
@@ -183,6 +209,47 @@ Tensor computeQuantizePerTensor(
+  ExprHandle out_qzero = LongImm::make(c10::get<int64_t>(inputs[3]));
+  Dtype dequant_dtype = kFloat;
+  Dtype out_dtype = outputType ? Dtype(*outputType) : QA.dtype();
+  std::vector<VarPtr> vars;
+  std::vector<ExprHandle> indices;
+  for (const auto& os : outputShape) {
+    auto var = alloc<Var>("", os.node()->dtype());
+    vars.push_back(var);
+    indices.push_back(VarHandle(var));
+  }
+  auto lhs = tensorOrConstant(inputs[0], indices);
+  auto rhs = tensorOrConstant(inputs[1], indices);
+  ExprHandle exprHandle = quant(
+      dequant(lhs, dequant_dtype, qa_scale, qa_zero) +
+          dequant(rhs, dequant_dtype, qb_scale, qb_zero),
+      out_dtype,
+      out_qscale,
+      out_qzero);
+  BufPtr buf = alloc<Buf>(
+      "quantized_add",
+      ExprHandleVectorToExprVector(outputShape),
+      out_dtype,
+      nullptr,
+      c10::nullopt,
+      out_qscale.node(),
+      out_qzero.node());
```
I suspect it would be more compact if we used `Compute` API to create the result `Tensor`.

ZolotukhinM reviewed (approved):
LGTM!

IvanKobzarev committed:
Update on "[nnc] tensorexpr for quantized::add"


Differential Revision: [D33238093](https://our.internmc.facebook.com/intern/diff/D33238093)

[ghstack-poisoned]

IvanKobzarev commented:
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33238093).
